src/topology/unit_test.rs

```diff
@@ -70,21 +103,43 @@ impl UnitTest {
         );
 
         for check in &self.checks {
-            if let Some(results) = results.get(&check.extract_from) {
-                let mut failed_conditions = Vec::new();
-                for (index, cond) in check.conditions.iter().enumerate() {
-                    if results.iter().find(|e| cond.check(e)).is_none() {
-                        failed_conditions.push(index);
-                    }
+            if let Some((inputs, outputs)) = results.get(&check.extract_from) {
+                if check.conditions.is_empty() {
+                    inspections.push(format!(
+                        "check transform '{}' payloads (JSON encoded):\n{}\n{}",
+                        check.extract_from,
+                        events_to_string("input", inputs),
+                        events_to_string("output", outputs),
+                    ));
+                    continue;
                 }
+                let failed_conditions = check
+                    .conditions
+                    .iter()
+                    .enumerate()
+                    .flat_map(|(i, cond)| {
+                        outputs
+                            .iter()
+                            .enumerate()
+                            .filter_map(|(j, e)| {
+                                cond.check_with_context(e).err().map(|err| {
+                                    if outputs.len() > 1 {
+                                        format!("condition[{}], payload[{}]: {}", i, j, err)
+                                    } else {
+                                        format!("condition[{}]: {}", i, err)
+                                    }
+                                })
+                            })
+                            .collect::<Vec<_>>()
+                    })
+                    .collect::<Vec<_>>();
                 if !failed_conditions.is_empty() {
-                    let event_strings: Vec<String> =
-                        results.iter().map(|e| event_to_string(e.clone())).collect();
                     errors.push(format!(
-                        "check transform '{}' failed conditions: [ {} ], payloads (encoded in JSON format):\n  {}",
+                        "check transform '{}' failed conditions:\n  {}\npayloads (JSON encoded):\n{}\n{}",
```

Comment:
Is it necessary to say this is JSON encoded? It seems like it might be obvious from looking at it.

Reply:
I think so, I'm trying to avoid a situation where a user is testing JSON parsing in their pipeline and is unable to rule out that the parsing has failed and we're displaying the unparsed contents of `message`.